utilities/src/with_std/dynamic_events.rs

```diff
@@ -0,0 +1,419 @@
+		self.0
+			.pointer("/phase/ApplyExtrinsic")
+			.and_then(|index| index.as_u64())
+			.filter(|index| *index == extrinsic_index)
+			.map(|_| self.event())
+	}
+
+	/// Returns the inner event and it's extrinsic index if the event was triggered during the
+	/// ApplyExtrinsic phase.
+	pub fn indexed_extrinsic_event(&self) -> Option<(u64, DynamicEvent)> {
+		self.0
+			.pointer("/phase/ApplyExtrinsic")
+			.and_then(|index| index.as_u64())
+			.map(|index| (index, self.event()))
+	}
+
+	/// Return a clone of the inner event.
+	pub fn event(&self) -> DynamicEvent {
+		DynamicEvent(
+			self.0
+				.pointer("/event")
+				.expect("existence of /event is ensured by constructor")
+				.clone(),
+		)
+	}
+
+	/// Unwrap the inner JSON value.
+	pub fn into_inner(self) -> JsonValue {
+		self.0
+	}
+}
+
+impl DynamicEvent {
+	/// If this event represents the outcome of an extrinsic, return the result.
+	pub fn extrinsic_outcome(&self) -> Option<Result<(), DispatchError>> {
+		if self.0.pointer("/System/ExtrinsicSuccess/dispatch_info").is_some() {
+			Some(Ok(()))
+		} else {
+			self.0
+				.pointer("/System/ExtrinsicFailed/dispatch_error")
+				.map(|dispatch_error_json| {
+					Err((|| {
+						Some(DispatchError::KnownModuleError {
+							pallet: dispatch_error_json["pallet"].as_str()?.to_string(),
+							name: dispatch_error_json["name"].as_str()?.to_string(),
+							error: dispatch_error_json["error"].as_str()?.to_string(),
+						})
+					})()
+					.unwrap_or_else(|| DispatchError::DispatchError(dispatch_error_json.clone())))
+				})
+		}
+	}
+
+	pub fn to_json(self) -> JsonValue {
+		self.0.clone()
+	}
+}
+
+type NamedVariantLookup<T> = BTreeMap<u8, (String, T)>;
+type ErrorInfo = NamedVariantLookup<Vec<String>>;
+
+/// A decoder for substrate events.
+///
+/// Events are decoded dynamically based on metadata provided on construction of this type.
+///
+/// For the System.ExtrinsicFailed event, the dispatch error is resolved to a structured
+/// human-readable error message with the following shape:
+///
+/// ```json
+/// {
+///     "pallet": "PalletName",
+///     "name": "ErrorName",
+///     "error": "Human readable error message"
+/// }
+/// ```
+#[derive(Debug)]
+pub struct EventDecoder {
+	events_type_id: u32,
+	types: PortableRegistry,
+	errors: NamedVariantLookup<ErrorInfo>,
+}
+
+impl EventDecoder {
+	/// Create a new event decoder.
+	///
+	/// The arguments to this constructor can be extracted from substrate metadata.
+	pub fn new(mut types: PortableRegistry, errors_type_id: u32) -> Option<Self> {
+		let errors = match types.resolve(errors_type_id).unwrap().type_def.clone() {
+			scale_info::TypeDef::Variant(runtime_error_type) => runtime_error_type
+				.variants
+				.into_iter()
+				.map(|pallet_error_type| {
+					(
+						pallet_error_type.index,
+						(pallet_error_type.name, {
+							let type_id = pallet_error_type
+								.fields
+								.first()
+								.expect("error variant has exactly one field")
```

Comment:
nit: technically this is asserting that we have *at least* one field. if we wanted to asset exactly one field, we could do a `.next()` and then check is_empty() or something.

Reply:
The fact it has exactly one field is what guarantees that first() is not None. ('at least one field' is implicit in the expect, 'exactly one field' is what explains why we assume there to be at least one... does that make sense?)

Also I'm not sure why i used first instead of next(). Maybe fields is not an iterator...

Reply:
> The fact it has exactly one field is what guarantees that first() is not None

I'm not sure what you mean. If there are two fields, then first() would be some, but there would be two fields and not exactly one?

Reply:
Maybe clearer: this code doesn't assert that we have exactly one field. It asserts that we have at last one field, and it does so confidently because we assume that there is exactly one field.

Reply:
https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.exactly_one